### PR TITLE
Shadowmonkers Typeshift

### PR DESCRIPTION
**Pull request by ayyaruq**
Depends on #509.

Migrate MNK to Typescript. Fists got kinda big change, everything else is mostly syntax and a bit of structure. Pretty much none of this works as an analysis in 5.0 but the code should build at least. My local git client thinks changes were ok but GitHub wants to treat stuff as deleted/added so derp.

### CODE REVIEW

File: src/parser/jobs/mnk/modules/GreasedLightning.tsx [hunk 1]
```diff
@@ -146,17 +154,18 @@ export default class GreasedLightning extends Module {
 		return events
 	}
 
-	_onGlGain(event) {
+	// This is fucked, idk why fflogs has a stack field in BuffEvent, our types don't
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/GreasedLightning.tsx]: wat

File: src/parser/jobs/mnk/modules/GreasedLightning.tsx [hunk 2]
```diff
@@ -299,7 +308,7 @@ export default class GreasedLightning extends Module {
 		// TODO: figure out how to make this graph at least 3x shorter in height
 
 		// Disabling magic numbers for the chart, 'cus it's a chart
-		/* eslint-disable no-magic-numbers */
+		/* tslint:disable no-magic-numbers */
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/GreasedLightning.tsx]: hot fucking damn i'm looking forward to nuking tslint

File: src/parser/jobs/mnk/modules/MnkAoE.tsx [hunk 3]
```diff
@@ -1,55 +1,54 @@
...
-		this.addHook('complete', this._onComplete)
+		this.addHook('complete', this.onComplete)
 	}
 
 	// TODO: figure out when player uses this for Silence effect, need to calculate interrupts on target
-	_onAotDDamage(event) {
-		if (event.hits.length < AOTD_MIN_TARGETS) {
+	private onAotDDamage(event: Event): void {
+		if (event.hasOwnProperty('hits') &&
+			(event as any).hits.length < AOTD_MIN_TARGETS) {
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/MnkAoE.tsx]: AoE event types (should) be on master now, can type the fn sig with those so you don't need this trash. Alternatively, type the fn sig with a `TODO`, and fix w/ later pr.

File: src/parser/jobs/mnk/modules/RiddleOfFire.tsx [hunk 4]
```diff
@@ -24,69 +26,73 @@ const ROF_GCD = {
...
+		this.addHook('complete', this.onComplete)
 	}
 
-	_onCast(event) {
-		const actionId = event.ability.guid
+	onCast(event: CastEvent): void {
+		const action = getDataBy(ACTIONS, 'id', event.ability.guid) as TODO // should be Action type
 
-		if (actionId === ACTIONS.RIDDLE_OF_FIRE.id) {
+		if (action.id === ACTIONS.RIDDLE_OF_FIRE.id) {
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/RiddleOfFire.tsx]: Technically speaking, `getDataBy<T>` has a return type of `T | undefined`, so you should be existence checking it.

File: src/parser/jobs/mnk/modules/RiddleOfFire.tsx [hunk 5]
```diff
@@ -146,7 +152,7 @@ export default class RiddleOfFire extends Module {
 		this._missedTks += 1 - tks.length
 	}
 
-	_formatGcdCount(count) {
+	private formatGcdCount(count: number): TODO {
```
Comment (ackwell) [inline src/parser/jobs/mnk/modules/RiddleOfFire.tsx]: Can this be `JSX.Element`?

File: src/parser/jobs/mnk/modules/BuffUptime.tsx [hunk 6]
```diff
@@ -1,94 +1,66 @@
...
+	protected init() {
 		// Hook all GCDs so we can count GCDs in buff windows
-		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('cast', {by: 'player'}, this.onCast)
 
-		// Apparently DK is a bit busted but let's use the debuff for now
-		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.UNKNOWN.id}, this._onDragonKick)
-		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this._onTwinSnakes)
+		this.addHook('applybuff', {to: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this.onGain)
+		this.addHook('refreshbuff', {to: 'player', abilityId: REFRESHERS}, this.onRefresh)
```
[the remark this answers is not in the document]
Comment (ayyaruq) [reply src/parser/jobs/mnk/modules/BuffUptime.tsx]: There's also Four Point Fury but that's for after new actions are added.

**Comment by ayyaruq** [comment]
Yep, but it felt weird to use tiered suggestion when there's only a single "yes or no" for it.

**Pull request merged**